File: Simulation/src/com/oberger/kruppelbotsimulation/function/WrappedPolyFunction.java
package com.oberger.kruppelbotsimulation.function;

import com.oberger.kruppelbotsimulation.util.IReadOnlyVector2;
import java.util.List;

/**
 * A decorator that wraps an {@link IPolyFunction} onto a given period (starting at x = 0).
 * {@link #getValue(float) } will return getValue( {@linkplain #mapOntoPeriod(float) }(x + offsetX)).
 * @author ole
 */
public class WrappedPolyFunction implements IPolyFunction {

    private IPolyFunction decorated = null;
    private float period;
    private float offsetX;
    
    public WrappedPolyFunction(IPolyFunction decorated, float period, float offsetX) {
        if (decorated == null) {
            throw new IllegalArgumentException(new NullPointerException("Passing null is not allowed."));
        }
        if (period <= 0) {
            throw new IllegalArgumentException("Period must be higher than zero but was " + period);
        }
        this.decorated = decorated;
        this.period = period;
        this.offsetX = offsetX;
    }

    @Override
    public float getValue(float x) {
        float mappedX = mapOntoPeriod(x + offsetX);
        
        return decorated.getValue(mappedX);
    }
    
    private float mapOntoPeriod(float x) {
        float mappedX = x;
        
        while (mappedX < 0) {
            mappedX += period;
        }
        mappedX %= period;
        
        return mappedX;
    }

    @Override
    public List<IReadOnlyVector2> getPolygons() {
        return decorated.getPolygons();
    }

    public float getPeriod() {
        return period;
    }
    
    public float getOffsetX() {
        return offsetX;
    }

    @Override
    public Interpolator getInterpolator() {
        return decorated.getInterpolator();
    }
    
}

File: Arduino/WrappedPolyFunction.cpp
#include "WrappedPolyFunction.h"

WrappedPolyFunction::WrappedPolyFunction(Vector2_t* polygons, const int polygonsLength, const unsigned int period, const unsigned int offsetX) : polygons(polygons), polygonsLength(polygonsLength), period(period), offsetX(offsetX)
{
}

WrappedPolyFunction::~WrappedPolyFunction()
{
	delete[] polygons;
}

int WrappedPolyFunction::getY(const int x)
{
	int mappedX = wrapOntoPeriod(x);
	int i;
	int interpolatedValue = 0;
	
	for (i = 0; i < polygonsLength - 1; i++)
	{
		Vector2_t p1 = polygons[i];
		Vector2_t p2 = polygons[i + 1];

		if (p1.x == mappedX)
		{
			interpolatedValue = p1.y;
		}
		else if (p2.x == mappedX)
		{
			interpolatedValue = p2.y;
		}
		else if (p1.x < mappedX && mappedX < p2.x)
		{
			interpolatedValue = interpolate(p1.x, p1.y, p2.x, p2.y, mappedX);
		}
	}

	return interpolatedValue;
}

int WrappedPolyFunction::wrapOntoPeriod(int x)
{
	int mappedX = x + offsetX;

	while (mappedX < 0)
	{
		mappedX += period;
	}
	mappedX %= period;

	return mappedX;
}

unsigned int WrappedPolyFunction::getPeriod()
{
	return period;
}

unsigned int WrappedPolyFunction::getOffsetX()
{
	return offsetX;
}
